### ISSUE TITLE
A dead symbolic link in a app/assets/images/ directory makes a rails app to fail loading

### ISSUE THREAD

**Issue by cmrd-senya** [jhass/js_image_paths#2]
Steps to reproduce:
1) Create a new Rails 4 app:

```
rails new blog
```

2) Add the following to Gemfile:

```
gem 'js_image_paths', '~> 0.0.2'
gem 'sprockets', '~> 2.12.4'
```

and `bundle update` then.
3) Execute

```
ln -s /dead_link app/assets/images/dead_link
bin/rails generate controller welcome index
bin/rails server
```

4) Check http://localhost:3000/welcome/index
You'll see an error message:
<img width="960" alt="2015-09-28 01 19 15" src="https://cloud.githubusercontent.com/assets/10187586/10125497/7d38b514-657f-11e5-8d14-571c634101c5.png">

**Comment by jhass** [comment]
Ah damn, it's coming from here? :D I'll look into it

**Comment by jhass** [comment]
So here's what happens.

I call `each_logical_path` https://github.com/rails/sprockets/blob/9937e2d3798350154d6d34aa62fb7a51d5d7d5a0/lib/sprockets/base.rb#L331 which calls `each_file` https://github.com/rails/sprockets/blob/9937e2d3798350154d6d34aa62fb7a51d5d7d5a0/lib/sprockets/base.rb#L319 which calls `each_entry` https://github.com/rails/sprockets/blob/9937e2d3798350154d6d34aa62fb7a51d5d7d5a0/lib/sprockets/base.rb#L298 and ends up on the problematic line: https://github.com/rails/sprockets/blob/9937e2d3798350154d6d34aa62fb7a51d5d7d5a0/lib/sprockets/base.rb#L307

`stat` returns `nil` if the file does not exist (and `File.exists` returns `false` for a dead symlink) https://github.com/sstephenson/hike/blob/f313ea1fd44caf2cae004fd45c67017b744bbb7a/lib/hike/index.rb#L91

Note that up to here no filter or anything was called. Now when `stat(path)` is reached with `path` pointing to the dead symlink, `stat` returns `nil` the chained `directory?` fails obviously.

Given we're behind by a major on both sprockets and hike and given I see no workaround this is probably a won't fix. Admittedly caused by using not really public API, but still. I wouldn't be surprised if it fixes itself as we're able to upgrade to newer versions of the involved dependencies.

**Comment by cmrd-senya** [comment]
Well yes, if we don't force sprockets to an old version, then it doesn't happen. So at least we know what's going on.

**Comment by jhass** [comment]
See also https://github.com/diaspora/diaspora/issues/6433

**Comment by SuperTux88** [comment]
I can't reproduce this anymore, neither with sprockets 3 and the released 0.1.1 version, nor with sprockets 4 (with the changes from #7 which are required to make sprockets 4 work). And sprockets 2 as used in this issue isn't supported anymore anyway. So I guess this can be closed.
